cache/cache.go

```diff
@@ -0,0 +1,190 @@
+package cache
+
+import (
+	"container/list"
+	"sync"
+)
+
+// Cache is a bounded-size in-memory cache of sized items with a configurable eviction policy
+type Cache interface {
+	// Get retrieves items from the cache by key.
+	// If an item for a particular key is not found, its position in the result will be nil.
+	Get(keys ...string) []Item
+
+	// Put adds an item to the cache.
+	Put(key string, item Item)
+
+	// Remove clears items with the given keys from the cache
+	Remove(keys ...string)
+
+	// Size returns the size of all items currently in the cache.
+	Size() uint64
+}
+
+// Item is an item in a cache
+type Item interface {
+	// Size returns the item's size, in bytes
+	Size() uint64
+}
+
+// A tuple tracking a cached item and a reference to its node in the eviction list
+type cached struct {
+	item    Item
+	element *list.Element
+}
+
+// Sets the provided list element on the cached item if it is not nil
+func (c *cached) setElementIfNotNil(element *list.Element) {
+	if element != nil {
+		c.element = element
+	}
+}
+
+// Private cache implementation
+type cache struct {
+	sync.Mutex                                  // Lock for synchronizing Get, Put, Remove
+	cap          uint64                         // Capacity bound
+	size         uint64                         // Cumulative size
+	items        map[string]*cached             // Map from keys to cached items
+	keyList      *list.List                     // List of cached items in order of increasing evictability
+	recordAdd    func(key string) *list.Element // Function called to indicate that an item with the given key was added
+	recordAccess func(key string) *list.Element // Function called to indicate that an item with the given key was accessed
+}
+
+// CacheOption configures a cache.
+type CacheOption func(*cache)
+
+// Capacity sets the fixed capacity bound on the cache, in bytes.
+// If not provided, default is 10MB.
+func Capacity(cap uint64) CacheOption {
+	return func(c *cache) {
+		c.cap = cap
+	}
+}
+
+// Policy is a cache eviction policy for use with the EvictionPolicy CacheOption.
+type Policy uint8
+
+const (
+	// LeastRecentlyAdded indicates a least-recently-added eviction policy.
+	LeastRecentlyAdded Policy = iota
+	// LeastRecentlyUsed indicates a least-recently-used eviction policy.
+	LeastRecentlyUsed
+)
+
+// EvictionPolicy sets the eviction policy to be used to make room for new items.
+// If not provided, default is LeastRecentlyUsed.
+func EvictionPolicy(policy Policy) CacheOption {
+	return func(c *cache) {
+		switch policy {
+		case LeastRecentlyAdded:
+			c.recordAccess = c.noop
+			c.recordAdd = c.record
+		case LeastRecentlyUsed:
+			c.recordAccess = c.record
+			c.recordAdd = c.noop
+		}
+	}
+}
+
+// New returns a cache with the requested options configured.
+// The cache consumes memory bounded by a fixed capacity,
+// plus tracking overhead linear in the number of items.
```

Comment:
If someone just wants to use the default options this might be the only docstring they read.  Might want to mention the 10MB default here.